plugins/quartertiler/Main.vala

```diff
@@ -0,0 +1,112 @@
+//
+//  Copyright (C) 2018 Peter Uithoven
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+namespace Gala.Plugins.QuarterTiler
+{
+	public class Main : Gala.Plugin
+	{
+		public override void initialize (Gala.WindowManager wm)
+		{
+			unowned Meta.Display display = wm.get_screen ().get_display ();
+
+			var settings = new GLib.Settings (Config.SCHEMA + ".keybindings");
+
+			display.add_keybinding ("tile-topleft", settings, Meta.KeyBindingFlags.NONE, (Meta.KeyHandlerFunc) on_initiate);
+			display.add_keybinding ("tile-topright", settings, Meta.KeyBindingFlags.NONE, (Meta.KeyHandlerFunc) on_initiate);
+			display.add_keybinding ("tile-bottomleft", settings, Meta.KeyBindingFlags.NONE, (Meta.KeyHandlerFunc) on_initiate);
+			display.add_keybinding ("tile-bottomright", settings, Meta.KeyBindingFlags.NONE, (Meta.KeyHandlerFunc) on_initiate);
+			display.add_keybinding ("tile-top", settings, Meta.KeyBindingFlags.NONE, (Meta.KeyHandlerFunc) on_initiate);
+			display.add_keybinding ("tile-bottom", settings, Meta.KeyBindingFlags.NONE, (Meta.KeyHandlerFunc) on_initiate);
+		}
+
+		[CCode (instance_pos = -1)]
+		void on_initiate (Meta.Display display, Meta.Screen screen,
+			Meta.Window? window, Clutter.KeyEvent event, Meta.KeyBinding binding)
+		{
+			unowned Meta.Window focused_window = display.get_focus_window ();
+			bool was_maximized_vertically = focused_window.maximized_vertically;
+			Meta.Rectangle prev_rect = focused_window.get_frame_rect ();
+
+			if (focused_window.maximized_vertically || focused_window.maximized_horizontally) {
+				focused_window.unmaximize (Meta.MaximizeFlags.BOTH);
+			}
+
+			if (focused_window.fullscreen) {
+				focused_window.unmake_fullscreen ();
+			}
+
+			Meta.Rectangle wa = focused_window.get_work_area_current_monitor ();
+			int x = wa.x;
+			int y = wa.y;
+			int width = wa.width / 2;
+			int height = wa.height / 2;
+			switch (binding.get_name ()) {
+				case "tile-topleft":
+				default:
+					break;
+				case "tile-topright":
+					x += width;
+					break;
+				case "tile-bottomleft":
+					y += height;
+					break;
+				case "tile-bottomright":
+					x += width;
+					y += height;
+					break;
+				case "tile-top":
+					if (!was_maximized_vertically) {
```

Comment:
This `if` logic means a window in the bottom-left corner cannot be moved to the top-left corner by pressing ctrl+meta+up. I think this is a bit of a missed opportunity.

Reply:
Conversely, maximized windows do react to ctrl+meta+down, and move to the bottom left corner. Maybe it would be more consistent if they occupied the bottom half of the screen? (And vice versa for up)

Reply:
But doing that opens up a lot of edge cases, for example for determining on which side the window is. Since I'm limiting myself to already tiled windows it's left when x is 0, it's on the right when it's not.

Reply:
`maximized_vertically` is true for tiled windows *and* maximized windows, which is why maximized windows can be shrunk into the bottom left corner using ctrl+shift+down right now.

I feel either the up/down shortcuts should be ignored in maximized windows, or they should do something that does not randomly favor the left over the right side (e.g. by using the full bottom half of the screen in this special case).

Reply:
The code would still be simpler that way: If the cases for "tile-top" and "tile-bottom" would leave the `x` and `width` of the window alone, and only adjust `y` and `height`, then ctrl+shift+up could be used to move from "left half" to "top left corner", etc.

Reply:
Good points, except for the fact that when I "unmaximize" a window it's position and size will be restored to a pre-tiled state. If I don't unmaximize I can't change the position and size. 

> I feel either the up/down shortcuts should be ignored in maximized windows, or they should do something that does not randomly favor the left over the right side (e.g. by using the full bottom half of the screen in this special case).

Good point! I'm now tiling to the full bottom or top of the screen when the window was maximized.